src/osd/pg_scrubber.cc

```diff
@@ -58,6 +72,1822 @@ ostream& operator<<(ostream& out, const requested_scrub_t& sf)
+
+	  e = m_pg->osd->store->queue_transaction(m_pg->ch, std::move(t));
+	  if (e != 0) {
+	    derr << __func__ << ": queue_transaction got " << cpp_strerror(e) << dendl;
+	  } else {
+	    std::unique_lock l{my_lock};
+	    my_cond.wait(l, [&done] { return done; });
+	  }
+	}
+      }
+    }
+  }
+}
+
+int PgScrubber::build_primary_map_chunk()
+{
+  auto ret = build_scrub_map_chunk(m_primary_scrubmap, m_primary_scrubmap_pos, m_start,
+				   m_end, m_is_deep);
+
+  if (ret == -EINPROGRESS)
+    m_osds->queue_for_scrub_resched(m_pg, Scrub::scrub_prio_t::high_priority);
+
+  return ret;
+}
+
+int PgScrubber::build_replica_map_chunk()
+{
+  dout(10) << __func__ << " epoch start: " << m_epoch_start << " ep q: " << m_epoch_queued
+	   << dendl;
+  dout(10) << __func__ << " deep: " << m_is_deep << dendl;
+
+  auto ret = build_scrub_map_chunk(replica_scrubmap, replica_scrubmap_pos, m_start, m_end,
+				   m_is_deep);
+
+  if (ret == 0) {
+
+    // finished!
+    // In case we restarted smaller chunk, clear old data
+
+    ScrubMap for_meta_scrub;
+    m_cleaned_meta_map.clear_from(m_start);
+    m_cleaned_meta_map.insert(replica_scrubmap);
+    clean_meta_map(for_meta_scrub);
+    _scan_snaps(for_meta_scrub);
+  }
+
+  // previous version used low priority here. Now switched to using the priority
+  // of the original message
+  if (ret == -EINPROGRESS)
+    requeue_replica(m_replica_request_priority);
+
+  return ret;
+}
+
+int PgScrubber::build_scrub_map_chunk(
+  ScrubMap& map, ScrubMapBuilder& pos, hobject_t start, hobject_t end, bool deep)
+{
+  dout(10) << __func__ << " [" << start << "," << end << ") "
+	   << " pos " << pos << " Deep: " << deep << dendl;
+
+  // start
+  while (pos.empty()) {
+
+    pos.deep = deep;
+    map.valid_through = m_pg->info.last_update;
+
+    // objects
+    vector<ghobject_t> rollback_obs;
+    pos.ret =
+      m_pg->get_pgbackend()->objects_list_range(start, end, &pos.ls, &rollback_obs);
+    dout(10) << __func__ << " while pos empty " << pos.ret << dendl;
+    if (pos.ret < 0) {
+      dout(5) << "objects_list_range error: " << pos.ret << dendl;
+      return pos.ret;
+    }
+    dout(10) << __func__ << " pos.ls.empty()? " << (pos.ls.empty() ? "+" : "-") << dendl;
+    if (pos.ls.empty()) {
+      break;
+    }
+    m_pg->_scan_rollback_obs(rollback_obs);
+    pos.pos = 0;
+    return -EINPROGRESS;
+  }
+
+  // scan objects
+  while (!pos.done()) {
+    int r = m_pg->get_pgbackend()->be_scan_list(map, pos);
+    dout(10) << __func__ << " be r " << r << dendl;
+    if (r == -EINPROGRESS) {
+      dout(8 /*20*/) << __func__ << " in progress" << dendl;
+      return r;
+    }
+  }
+
+  // finish
+  dout(8 /*20*/) << __func__ << " finishing" << dendl;
+  ceph_assert(pos.done());
+  m_pg->_repair_oinfo_oid(map);
+
+  dout(8 /*20*/) << __func__ << " done, got " << map.objects.size() << " items" << dendl;
+  return 0;
+}
+
+/**
+ * \todo describe what we are doing here
```

Comment:
@ronen-fr Caller passes an empty ScrubMap and gets back a map of objects suitable for snapshot validation.    That means that first item must be a head object and the last object must be a head object with no snapshots or the last snapshot of an object.

The data in m_cleaned_meta_map  is the left over partial items that need to be completed before they can be processed.

Reply:
I had forgotten that the maps go in reverse order.  Snapshots precede the head object, which is why the code works on the map backwards.